cpp/src/arrow/util/reflection_internal.h

```diff
@@ -0,0 +1,192 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <array>
+#include <string>
+#include <tuple>
+#include <utility>
+
+namespace arrow {
+namespace internal {
+
+namespace detail {
+
+#ifdef _MSC_VER
+#define ARROW_PRETTY_FUNCTION __FUNCSIG__
+#else
+#define ARROW_PRETTY_FUNCTION __PRETTY_FUNCTION__
+#endif
+
+template <typename T>
+const char* raw() {
```

Comment:
`raw()` and `raw_sizeof()` are undescriptive names.  Consider renaming to `pretty_function_name()` and `prettty_function_name_size()`.  

Is there a reason why these functions are not `constexpr`?

Reply:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66639
Older GCC doesn't think __PRETTY_FUNCTION__ is constexpr. Shame, too: with constexpr it's possible to ensure the sliced-away portions of __PRETTY_FUNCTION__ don't make it into the object file